ENH: Restructure libcpp.vector and cover API with tests

zyv opened the pull request:
In particular, add C++11 declarations, more const-correctness.

I have added a new file for testing C++11 methods, I think this can be later used to add more tests for container methods and algorithms, which have been added with C++11 and were not tested before, because otherwise these tests will fail on compilers where the default mode is still not C++11.

zyv commented:
Will rework as requested in #488.

scoder commented:
Any update on this?

zyv commented:
I resolved the conflicts with the current master, and manually went through the declarations one by one, picking mine first and then trying to see if it has since been adopted in the master branch. This declaration list is somewhat smaller than the string one, so I actually managed to do this in a finite amount of time.

I still prefer my approach in 0a51a899d691be6ab62d86429e2bd24b69f7c8a7, because in my opinion this header is completely unmaintainable, but I understand that any attempt at a major restructuring will probably end up in limbo unless (1) we agree on a good restructuring strategy, and (2) there is some way to validate the results.

It's the same story as #488, though, so it makes sense to get consensus there first, and maybe this can just be accepted as is (without restructuring) now?

@scoder , @da-woods what do you think?

zyv commented on Cython/Includes/libcpp/vector.pxd:
```diff
@@ -113,20 +113,20 @@ cdef extern from "<vector>" namespace "std" nogil:
+        bint operator>=(const vector&, const vector&)
         void assign(size_type, const T&)
-        void assign[InputIt](InputIt, InputIt) except +
+        void assign(iterator, iterator) except +
```
Here I noticed that this would break Cython code that previously passed `const_iterator` to `assign`, because Cython is not smart enough about this.

What do you think the strategy should be?

* Obviously, adding definitions with all possible combinations of `const_iterator` and `iterator` is not feasible.
* Templating as done here leaves the type checking to the compiler, which is not ideal.

Should I go back and add a test to check that `const_iterator` is accepted here, so it doesn't break in the future?

What about other places where only non-const iterators were accepted, like `erase`, `insert`?

* Should they be templated?
* Should they be changed to `const_iterator` (possibly breaking older code)?
* Should they be left alone?

da-woods replied:
> Templating as done here leaves the type checking to the compiler, which is not ideal.

I'm somewhat relaxed about passing the type checking to the C++ compiler in general. C++ is complicated and Cython will never end up understanding it as well as the C++ compiler.

To me the templated version looks like the right definition (i.e. it matches what the standard says) so we should probably just do that. Because in principle it's any input iterator it takes rather than just vector iterators, right?

zyv replied:
> Because in principle it's any input iterator it takes rather than just vector iterators, right?

Well, I've actually confounded two problems.

1. Some overloads are defined as templates, without any further type restrictions:
  ```
  template<class InputIterator>
  basic_string& assign(InputIterator first, InputIterator last);
  ```
  I agree with you here. I would template it in Cython without a second thought.

2. Some overloads are defined as `const_iterator`, but obviously accept `iterator` as well. Here I'm really not sure what to do (see above):
    1. Leave it as it was defined for whatever reason (?!) - e.g. `const_iterator` just wasn't declared.
    2. Switch everything to `const_iterator` that can accept it, possibly breaking user code
    3. Add templating.

da-woods replied:
For the `const_iterator` thing: I think adding templating may cause problems if the function isn't a template function (because Cython insists on generating code like `s.func<SomeIteratorType>(first, second);` which is obviously invalid if `func` isn't a template).

The other option which can work is to write the function as `func(...)`, which just says "the arguments are completely up to the C++ compiler to sort out". I don't feel too bad about doing this where necessary, but if you do that then Cython can't even help you with the number of arguments. And you often have to define that as the only overload, otherwise Cython treats it as ambiguous.

It'd probably help you could could point me at a function where the `iterator`/`const_iterator` is causing problems (since `assign` possibly isn't the best example).

zyv replied:
Yes, let's look at `erase()`. It is now defined as

```cpp
iterator erase(const_iterator position);
iterator erase(const_iterator first, const_iterator last);
```

Up to C++11, it was defined as

```cpp
iterator erase(iterator position);
iterator erase(iterator first, iterator last);
```

which is probably where the Cython declaration comes from.

If Cython declaration is changed to `const_iterator`, old user code won't compile anymore because Cython won't automatically cast it. If you would leave it as `iterator`, it won't take `const_iterator` :(

Similar problem with `insert`, `emplace` (here it was just defined with `const_iterator` as an argument, because it didn't even exist before C++). The same goes for the constructor as well.

da-woods replied:
Thanks for the clarification - that makes sense.

My feeling is that templates won't work here because of Cython's code generation. I think the most robust way would be to define overloads for both `iterator` and `const_iterator` (but the 2 mixed versions because that seems excessive, and I'm sure people can work it out themselves). That seems reasonable to me in that both definitions have existed in the C++ standard at various times.

The other option might be to tell Cython that `iterator` inherits from `const_iterator`. I know that's a lie. But it might catch the idea that you can always pass an `iterator` where it expects a `const_iterator`. I'd be a little nervous that this might have unintended consequences. Especially since it's old code that lots of people are already using. So it might be best not to do this.

zyv replied:
> My feeling is that templates won't work here because of Cython's code generation. I think the most robust way would be to define overloads for both `iterator` and `const_iterator` (but the 2 mixed versions because that seems excessive, and I'm sure people can work it out themselves). That seems reasonable to me in that both definitions have existed in the C++ standard at various times.

I've tried templates and they've worked (at least for my limited testing), but it doesn't feel right to me, which is why I brought this up and wanted to get your opinion on it:

```diff
diff --git a/Cython/Includes/libcpp/vector.pxd b/Cython/Includes/libcpp/vector.pxd
index 49444a636..8a2399328 100644
--- a/Cython/Includes/libcpp/vector.pxd
+++ b/Cython/Includes/libcpp/vector.pxd
@@ -139,8 +139,8 @@ cdef extern from "<vector>" namespace "std" nogil:
         iterator end()
         const_iterator const_end "end"()
         const_iterator cend()
-        iterator erase(iterator) except +
-        iterator erase(iterator, iterator) except +
+        iterator erase[InputIt](InputIt) except +
+        iterator erase[InputIt](InputIt, InputIt) except +
         T& front()
         iterator insert(iterator, const T&) except +
         iterator insert(iterator, size_type, const T&) except +
diff --git a/tests/run/cpp_stl_vector.pyx b/tests/run/cpp_stl_vector.pyx
index 18fffddbd..de7d352e2 100644
--- a/tests/run/cpp_stl_vector.pyx
+++ b/tests/run/cpp_stl_vector.pyx
@@ -63,7 +63,7 @@ def test_modifiers():
     v2.insert(v2.end(), v1.begin(), v1.end())
     v1.clear()
     v2.pop_back()
-    v2.erase(v2.begin())
+    v2.erase(v2.cbegin())
     v2.erase(v2.begin(), v2.end())
     return v1.size(), v2.size()

```

> The other option might be to tell Cython that `iterator` inherits from `const_iterator`. I know that's a lie. But it might catch the idea that you can always pass an `iterator` where it expects a `const_iterator`. I'd be a little nervous that this might have unintended consequences. Especially since it's old code that lots of people are already using. So it might be best not to do this.

Yes, this is exactly how it was done when I did my rewrite (0a51a899d691be6ab62d86429e2bd24b69f7c8a7). However, in 169876872f3cb6198971a1db07e5b8a9d12b3dac it has been changed to the current definitions to make everything much stricter/type-safe.

So, if I understand you correctly, you would prefer to ship something like this, and that's what you would accept:

```diff
diff --git a/Cython/Includes/libcpp/vector.pxd b/Cython/Includes/libcpp/vector.pxd
index 49444a636..d853270a2 100644
--- a/Cython/Includes/libcpp/vector.pxd
+++ b/Cython/Includes/libcpp/vector.pxd
@@ -141,6 +141,8 @@ cdef extern from "<vector>" namespace "std" nogil:
         const_iterator cend()
         iterator erase(iterator) except +
         iterator erase(iterator, iterator) except +
+        iterator erase(const_iterator) except +
+        iterator erase(const_iterator, const_iterator) except +
         T& front()
         iterator insert(iterator, const T&) except +
         iterator insert(iterator, size_type, const T&) except +
```

If that's the case, that's fine with me.

da-woods replied:
I've double checked the code generation, and it looks like the generated code spells out the template types for free functions but not for methods. (I'm not sure why exactly). So the templated version is fine I think. What I was mostly worried about is the generated code reading `v.erase<std::vector<double>::iterator>(it1, it2)`; but it doesn't do this so no problem

zyv commented on Cython/Includes/libcpp/vector.pxd:
```diff
@@ -113,20 +113,20 @@ cdef extern from "<vector>" namespace "std" nogil:
-        vector(size_type, T&) except +
-        #vector[InputIt](InputIt, InputIt)
+        vector(size_type, const T&) except +
+        vector(iterator, iterator) except +
```
TODO: if templating - add fake type. Decision pending.

da-woods replied:
This should be templated I think because the constructor is for any input iterators rather than just the std::vector ones

zyv commented:
I have now done the following:

* Rebased the tests to the latest master
* Removed my reordering, leaving only `const` / `except +` changes
* Added a new constructor with a comment on how I made it work.
 
It seems that the rest of the definitions I added back then have since been integrated into the master, or at least my tests are passing.

Codespell errors are not mine, should they be fixed anyway? As far as I'm concerned, this is ready for merging.

Let me know if there is still interest in reordering according to the standard (in two steps, first order changes only, then additions if necessary).

da-woods commented:
> Codespell errors are not mine, should they be fixed anyway?

No quite sure why they're popping up in this PR only, but ignore them.

h-vetinari commented:
Wow, seeing a PR from 9 years ago merged is a big surprise 😅 Congratulations! 🥳
